**START**

Load in pandas and sqlalchemy's create_engine & text
Then connect to Postgres and create the engine

In [1]:
from sqlalchemy import create_engine, text
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt
import folium
from folium.plugins import MarkerCluster
database_name = 'scooters'
connection_string = f"postgresql://postgres:postgres@localhost:5433/{database_name}"
engine = create_engine(connection_string)

**EDA**

In [2]:
count_all_rows = '''
(SELECT
    'scooters' table,
    COUNT(*)
FROM scooters)
UNION
(SELECT
    'trips' table,
    COUNT(*)
FROM trips);
'''

with engine.connect() as connection:
    counts = pd.read_sql(text(count_all_rows), con = connection)

counts

table     count
0  scooters  73414043
1     trips    565522

In [3]:
find_nulls_scooters = '''
SELECT *
FROM scooters
WHERE NOT(scooters IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_scooters = pd.read_sql(text(find_nulls_scooters), con = connection)

find_nulls_trips = '''
SELECT *
FROM trips
WHERE NOT(trips IS NOT NULL);
'''

with engine.connect() as connection:
    nulls_trips = pd.read_sql(text(find_nulls_trips), con = connection)

nulls_scooters.info()
nulls_trips.info()

nulls_scooters

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 770 entries, 0 to 769
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   pubdatetime  770 non-null    datetime64[ns]
 1   latitude     770 non-null    float64       
 2   longitude    770 non-null    float64       
 3   sumdid       770 non-null    object        
 4   sumdtype     770 non-null    object        
 5   chargelevel  0 non-null      object        
 6   sumdgroup    770 non-null    object        
 7   costpermin   770 non-null    float64       
 8   companyname  770 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(5)
memory usage: 54.3+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 0 entries
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   pubtimestamp    0 non-null      object
 1   companyname     0 non-null      object
 2   triprecordn

pubdatetime   latitude  longitude  \
0   2019-06-15 23:20:36  36.158512 -86.780570   
1   2019-06-15 23:25:37  36.158560 -86.780545   
2   2019-06-15 23:30:37  36.158646 -86.780541   
3   2019-06-15 23:35:37  36.158682 -86.780519   
4   2019-06-15 23:40:37  36.158683 -86.780522   
..                  ...        ...        ...   
765 2019-06-14 18:39:33  36.150280 -86.813120   
766 2019-06-14 18:44:33  36.150258 -86.813127   
767 2019-06-14 19:04:34  36.150271 -86.813111   
768 2019-06-14 19:09:34  36.150271 -86.813111   
769 2019-06-15 20:10:30  36.158659 -86.780577   

                                           sumdid sumdtype chargelevel  \
0    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
1    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
2    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
3    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
4    Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   
..                                            ...      ...         ...   
765  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
766  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
767  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
768  Powered-72a04621-3a01-05f2-3c9b-abae6d7387ce  Powered        None   
769  Powered-751e19ec-b697-a51f-04ea-0bf9a2e125c9  Powered        None   

    sumdgroup  costpermin companyname  
0     Scooter        0.15        Bolt  
1     Scooter        0.15        Bolt  
2     Scooter        0.15        Bolt  
3     Scooter        0.15        Bolt  
4     Scooter        0.15        Bolt  
..        ...         ...         ...  
765   Scooter        0.15        Bolt  
766   Scooter        0.15        Bolt  
767   Scooter        0.15        Bolt  
768   Scooter        0.15        Bolt  
769   Scooter        0.15        Bolt  

[770 rows x 9 columns]

While the .info() part of the output is a bit counter-intuitave, it shows that there are 770 null values in the scooters table and none in the trips table.

All of the null values are in the chargelevel column, and looking at the full output, they all belong to Bolt and Spin.

In [4]:
date_range = '''
(SELECT
    'scooters' table,
    MIN(pubdatetime) begin,
    MAX(pubdatetime) end
FROM scooters)
UNION
(SELECT
    'trips' table,
    MIN(pubtimestamp) begin,
    MAX(pubtimestamp) end
FROM trips);
'''

with engine.connect() as connection:
    dates = pd.read_sql(text(date_range), con = connection)

dates

table                   begin                 end
0  scooters 2019-05-01 00:01:41.247 2019-07-31 23:59:57
1     trips 2019-05-01 00:00:55.423 2019-08-01 07:04:00

In [5]:
aug_first = '''
SELECT *
FROM trips
WHERE enddate > '2019-07-31'
LIMIT 100;
'''

with engine.connect() as connection:
    late = pd.read_sql(text(aug_first), con = connection)

late

pubtimestamp companyname triprecordnum  \
0  2019-08-01 00:01:02.110        Lyft          LFT2   
1  2019-08-01 00:00:06.593        Bird       BRD1816   
2  2019-08-01 00:00:06.593        Bird       BRD1819   
3  2019-08-01 00:00:06.593        Bird       BRD1818   
4  2019-08-01 00:00:06.593        Bird       BRD1820   
..                     ...         ...           ...   
73 2019-08-01 04:53:48.000        JUMP          JMP2   
74 2019-08-01 04:53:48.000        JUMP          JMP1   
75 2019-08-01 07:04:00.000        JUMP         JMP35   
76 2019-08-01 07:04:00.000        JUMP         JMP34   
77 2019-08-01 00:03:30.973        Lyft         LFT12   

                                         sumdid  tripduration  tripdistance  \
0                                 Powered318477      3.226600     790.68244   
1                                  PoweredYSEGE      7.000000    1968.50400   
2                                  PoweredYTT8X      8.000000       0.00000   
3                                  Powered799T4      7.000000       0.00000   
4                                  PoweredJF4AU     12.000000       0.00000   
..                                          ...           ...           ...   
73  Powereda4712099-5e55-5332-996c-d6e5c910535d      3.154917    1320.00000   
74  Poweredb8a3a269-d1ca-571f-9b2f-89b7399b5537     18.440650    7920.00000   
75  Poweredf077a919-d569-5e70-8ca7-71d179ffacf9    142.345610   20433.60000   
76  Powered784fcc92-bd1a-5f6a-8314-b02aa21b4bfa    209.409650   14889.60000   
77                                Powered599249      4.203000     790.68244   

     startdate        starttime     enddate          endtime  startlatitude  \
0   2019-07-31  23:57:48.396666  2019-08-01  00:01:01.993333      36.157490   
1   2019-07-31  23:54:46.096666  2019-08-01  00:01:29.436666      36.151100   
2   2019-07-31  23:52:40.086666  2019-08-01  00:00:41.923333      36.146100   
3   2019-07-31  23:53:17.616666  2019-08-01  00:00:27.170000      36.158100   
4   2019-07-31  23:48:24.240000  2019-08-01  00:00:16.660000      36.161200   
..         ...              ...         ...              ...            ...   
73  2019-07-31         23:57:19  2019-08-01         00:00:28      36.153015   
74  2019-07-31         23:41:52  2019-08-01         00:00:18      36.153687   
75  2019-07-31         23:26:15  2019-08-01         01:48:35      36.155735   
76  2019-07-31         22:19:08  2019-08-01         01:48:33      36.160050   
77  2019-07-31  23:59:18.720000  2019-08-01  00:03:30.900000      36.164790   

    startlongitude  endlatitude  endlongitude  \
0       -86.777140    36.157950    -86.776030   
1       -86.783300    36.152000    -86.791100   
2       -86.799500    36.151800    -86.797300   
3       -86.769200    36.158100    -86.769200   
4       -86.770900    36.162500    -86.774300   
..             ...          ...           ...   
73      -86.783585    36.152843    -86.779594   
74      -86.784580    36.164158    -86.776900   
75      -86.775185    36.173904    -86.785450   
76      -86.776720    36.163410    -86.782100   
77      -86.779840    36.165800    -86.777830   

                                            triproute               create_dt  
0   [(36.15749, -86.77714), (36.15748, -86.7772), ... 2019-08-02 10:27:50.310  
1   [(36.151103, -86.783327), (36.150872, -86.7831... 2019-08-02 05:30:21.230  
2   [(36.14615, -86.79952), (36.146129, -86.799557... 2019-08-02 05:30:21.340  
3                           [(36.158103, -86.769174)] 2019-08-02 05:30:21.290  
4   [(36.161257, -86.770772), (36.161298, -86.7707... 2019-08-02 05:30:21.373  
..                                                ...                     ...  
73  [('36.153016', '-86.783586'), ('36.153016', '-... 2019-08-02 08:21:26.773  
74  [('36.153685', '-86.784578'), ('36.153685', '-... 2019-08-02 08:21:26.803  
75  [('36.155736', '-86.775181'), ('36.155531', '-... 2019-08-02 08:24:21.967  
76  [('36.160051', '-86.77672'), ('36.160051', '-8... 

Both tables contain 3 months of data and both begin on May 1st, but the trips table ends a day after the scooters table. Looking at why, the trips all began before midnight on July 31st.

In [6]:
long_trips = '''
SELECT 
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration > 1440
GROUP BY companyname;
'''

with engine.connect() as connection:
    long = pd.read_sql(text(long_trips), con = connection)

long_trips

companyname  count
0  Bolt Mobility   6908
1           Lyft      2
2           SPIN     28

In [7]:
short_trips = '''
SELECT
    companyname,
    COUNT(*)
FROM trips
WHERE tripduration < 1
    AND tripdistance <= 0
GROUP BY companyname;
'''

with engine.connect() as connection:
    short = pd.read_sql(text(short_trips), con = connection)

short

companyname  count
0        Bird   3928
1        Lime    377
2        Lyft   3405

This data was supposed to have been cleaned before being submitted to the city, which includes stripping out all trips shorter than a minute or longer than 24 hours.

There are ~7,000 trips longer than 24 hours and over 9,000 trips shorter than one minute.

In [8]:
available_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM scooters
GROUP BY companyname
'''

with engine.connect() as connection:
    available = pd.read_sql(text(available_scooters), con = connection)

active_scooters = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) scooters
FROM trips
GROUP BY companyname
'''

with engine.connect() as connection:
    active = pd.read_sql(text(active_scooters), con = connection)

available

company  scooters
0    Bird      3860
1    Bolt       360
2  Gotcha       224
3    Jump      1210
4    Lime      1824
5    Lyft      1735
6    Spin       805

In [9]:
active

company  scooters
0           Bird      3766
1  Bolt Mobility       356
2         Gotcha       166
3           JUMP       450
4           Lime      1788
5           Lyft      1725
6           SPIN       754

In [10]:
availability = '''
SELECT
    companyname company,
    COUNT(DISTINCT sumdid) total_scooters
FROM scooters
WHERE sumdid NOT IN
    (SELECT
        DISTINCT sumdid
    FROM trips)
GROUP BY companyname;
'''

with engine.connect() as connection:
    unavailable = pd.read_sql(text(availability), con = connection)
    
unavailable

In [ ]:
daily_use = '''
SELECT 
    DISTINCT sumdid,
    companyname company,
    ROUND(AVG(COUNT(sumdid)) OVER (PARTITION BY sumdid, DATE(pubtimestamp)),2) avg_daily_usage
FROM trips
GROUP BY sumdid, company, pubtimestamp
'''

with engine.connect() as connection:
    usage = pd.read_sql(text(daily_use), con = connection)

multiple_uses = usage[usage['avg_daily_usage'] > 1.00]
multiple_uses

Filter trips table:
- Remove trips under a minute
- Remove all zero-distance trips
- Remove unreasonable long trips
    - These scooters have an average top speed of 15 mph, and generally have a range of about 50 miles.
    - That said, the batteries should last on average ~3.5 hours

In [6]:
trips_clean = '''
SELECT *
FROM trips
WHERE tripduration > 1.0
    AND tripduration < 200.0
    AND tripdistance > 0
    AND tripdistance < 264000
'''

with engine.connect() as connection:
    trips = pd.read_sql(text(trips_clean), con = connection)
    
trips.describe()

trips.head()

pubtimestamp companyname triprecordnum          sumdid  \
0 2019-06-02 20:58:41.033        SPIN        SPI886  Powered3745490   
1 2019-06-02 20:58:41.033        SPIN        SPI877  Powered8106751   
2 2019-06-02 20:58:41.033        SPIN        SPI879  Powered6903219   
3 2019-06-02 20:58:41.033        SPIN        SPI884  Powered6695098   
4 2019-06-02 20:58:41.033        SPIN        SPI863  Powered6228267   

   tripduration  tripdistance   startdate starttime     enddate   endtime  \
0           4.0     793.96328  2019-06-02  20:54:29  2019-06-02  20:57:54   
1          13.0     370.73492  2019-06-02  20:45:48  2019-06-02  20:58:31   
2          13.0     351.04988  2019-06-02  20:46:25  2019-06-02  20:58:37   
3           5.0     131.23360  2019-06-02  20:52:57  2019-06-02  20:57:11   
4          21.0    8674.54096  2019-06-02  20:34:10  2019-06-02  20:54:24   

   startlatitude  startlongitude  endlatitude  endlongitude  \
0      36.161966      -86.775011    36.163276    -86.777170   
1      36.152710      -86.785055    36.152981    -86.783840   
2      36.152711      -86.785046    36.152987    -86.783894   
3      36.161227      -86.777061    36.161449    -86.776705   
4      36.176285      -86.785433    36.152841    -86.790189   

                                           triproute               create_dt  
0  [(36.163276406038705, -86.7771697138313), (36.... 2019-06-03 09:05:38.947  
1  [(36.1529805561506, -86.78383959028197), (36.1... 2019-06-03 09:05:38.980  
2  [(36.152987202313604, -86.78389426748939), (36... 2019-06-03 09:05:39.010  
3  [(36.1614456, -86.7767024), (36.1614472, -86.7... 2019-06-03 09:05:38.880  
4  [(36.15284125996444, -86.79018921571804), (36.... 2019-06-03 09:05:38.753

**MAPPING**

In [7]:
zipcodes = gpd.read_file('../data/zipcodes.geojson')

zipcodes = zipcodes[['zip', 'po_name', 'geometry']]
# Create separate tables for start & end location data
scooter_start = trips[['companyname', 'triprecordnum', 'sumdid', 'startlongitude', 'startlatitude']]
scooter_start_geo = gpd.GeoDataFrame(scooter_start, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_start.startlongitude, scooter_start.startlatitude))
scooter_end = trips[['companyname', 'triprecordnum', 'sumdid', 'endlongitude', 'endlatitude']]
scooter_end_geo = gpd.GeoDataFrame(scooter_end, crs= zipcodes.crs, geometry=gpd.points_from_xy(scooter_end.endlongitude, scooter_end.endlatitude))

In [8]:
# Combine start and end locations by zip code
starting_zip = gpd.sjoin(scooter_start_geo, zipcodes, predicate='within')
ending_zip = gpd.sjoin(scooter_end_geo, zipcodes, predicate='within')

In [9]:
starting_zip.head()

companyname triprecordnum          sumdid  startlongitude  startlatitude  \
0         SPIN        SPI886  Powered3745490      -86.775011      36.161966   
3         SPIN        SPI884  Powered6695098      -86.777061      36.161227   
11        SPIN        SPI894  Powered8683450      -86.776591      36.163797   
13        SPIN        SPI888  Powered5330088      -86.774547      36.159697   
14        SPIN        SPI893  Powered8683450      -86.776228      36.163546   

                      geometry  index_right    zip    po_name  
0   POINT (-86.77501 36.16197)           36  37201  NASHVILLE  
3   POINT (-86.77706 36.16123)           36  37201  NASHVILLE  
11  POINT (-86.77659 36.16380)           36  37201  NASHVILLE  
13  POINT (-86.77455 36.15970)           36  37201  NASHVILLE  
14  POINT (-86.77623 36.16355)           36  37201  NASHVILLE

In [10]:
ending_zip.head()

companyname triprecordnum          sumdid  endlongitude  endlatitude  \
0         SPIN        SPI886  Powered3745490    -86.777170    36.163276   
3         SPIN        SPI884  Powered6695098    -86.776705    36.161449   
11        SPIN        SPI894  Powered8683450    -86.776694    36.163699   
12        SPIN        SPI868  Powered5743469    -86.776375    36.160219   
14        SPIN        SPI893  Powered8683450    -86.776350    36.163440   

                      geometry  index_right    zip    po_name  
0   POINT (-86.77717 36.16328)           36  37201  NASHVILLE  
3   POINT (-86.77671 36.16145)           36  37201  NASHVILLE  
11  POINT (-86.77669 36.16370)           36  37201  NASHVILLE  
12  POINT (-86.77637 36.16022)           36  37201  NASHVILLE  
14  POINT (-86.77635 36.16344)           36  37201  NASHVILLE

In [28]:
start_by_zip = pd.DataFrame(starting_zip['zip'].value_counts())
end_by_zip = pd.DataFrame(ending_zip['zip'].value_counts())
delta = pd.concat([start_by_zip[['company', 'x']], df2[['company', 'x']]])
delta = pd.pivot_table(all_zip, index=['zip'], columns=['start'], aggfunc={'values': 'sum'}).reset_index()

KeyError: 'start'

In [ ]:
# Get centroid of Davidson County
#county = gpd.read_file('../data/Davidson County Border (GIS).geojson')
#county.geometry.centroid
# Use ID #0
#center = county.geometry.centroid[0]
#map_center = [center.y, center.x]

# Draw map of starting positions for each trip
#scooter_start_map = folium.Map(location = map_center, zoom_start = 10)
# Add marker cluster
#marker_cluster = MarkerCluster().add_to(scooter_start_map)
# Add zipcodes to map
#folium.GeoJson(zipcodes).add_to(scooter_start_map)
# Use a for loop to add projects
#for row_index, row_values in starting_zip.iterrows():
#    loc = [row_values['startlatitude'], row_values['startlongitude']]
#    pop = str(row_values['companyname'])
#    icon=folium.Icon(color="blue",icon="exclamation-triangle", prefix='fa')
#    
#    marker = folium.Marker(
#        location = loc, 
#        popup = pop,
#    icon = icon) 
#    
#    marker.add_to(marker_cluster)
#
#scooter_start_map.save('../maps/starting_location.html')
# Display the map
#scooter_start_map